src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double resizeCoefficient = 1.5;
+    private Object[] elementData;
+    private int numberOfObjects;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        numberOfObjects = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapasity();
+        elementData[numberOfObjects] = value;
+        numberOfObjects++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > numberOfObjects) {
+            throw new ArrayListIndexOutOfBoundsException("Incorrect index");
+        }
+        ensureCapasity();
+        addElementInsideData(value, index);
+    }
 
+    private void addElementInsideData(T value, int index) {
+        Object[] tempFirstPart = Arrays.copyOfRange(elementData, 0, index);
+        Object[] tempSecondPart = Arrays.copyOfRange(elementData, index, elementData.length);
+
+        int counterForSecondArray = 0;
+        for (int i = 0; i < elementData.length; i++) {
+            if (i == index) {
+                numberOfObjects++;
+                elementData[i] = value;
+                continue;
+            }
+            if (i < index) {
+                elementData[i] = tempFirstPart[i];
+                continue;
+            }
+            if (i > index) {
+                elementData[i] = tempSecondPart[counterForSecondArray];
+                counterForSecondArray++;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+
+        Object[] tempFirstPart = Arrays.copyOfRange(elementData, 0, index);
+        Object[] tempSecondPart = Arrays.copyOfRange(elementData, index + 1, elementData.length);
+        numberOfObjects--;
+        Object result = elementData[index];
+
+        System.arraycopy(tempFirstPart, 0, elementData,
+                0, tempFirstPart.length);
+        System.arraycopy(tempSecondPart, 0, elementData,
+                tempFirstPart.length, tempSecondPart.length);
```

Comment:
the same here. I show you how you can simplify logic in the method ```addElementInsideData()```. You can use something similar here